docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,315 @@
+authors:
+- "@seshachalam-yv"
+- "@renormalize"
+- "@ishan16696"
+reviewers:
+- "@unmarshall"
+---
+
+# DEP-06: Immutable ETCD Backups
+
+## Summary
+
+This proposal aims to enhance the reliability and integrity of ETCD backups created by `etcd-backup-restore` in ETCD clusters managed by `etcd-druid`, by introducing immutable backups. By leveraging cloud provider features that support a write-once-read-many (WORM) model, unauthorized modifications to backup data are prevented, ensuring that backups remain intact and accessible for restoration.
+
+The proposed solution relies on `etcd-druid` to manage ETCD backups and handle hibernation processes effectively. It leverages one of the suggested approaches to ensure backups remain immutable over extended periods. It is important to note that using `etcd-backup-restore` standalone may not be sufficient to achieve this functionality end-to-end, as the immutability handling (with respect to hibernation) is specifically managed within `etcd-druid`.
+
+## Motivation
+
+Ensuring the integrity and availability of ETCD backups is crucial for the ability to restore an ETCD cluster when it has become non-functional or inoperable. Making the backups immutable protects against any unintended or malicious modifications post-creation, thereby enhancing the overall security posture.
+
+### Goals
+
+- Implement immutable backup support for ETCD clusters.
+- Secure backup data against unintended or unauthorized modifications after creation.
+- Implement changes required in `etcd-backup-restore` and `etcd-druid` to support this proposal.
+
+### Non-Goals
+
+- Implementing object-level immutability policies at this stage.
+- Supporting immutable backups on storage providers that do not offer immutability features (e.g., OpenStack Swift).
+
+## Proposal
+
+### Overview
+
+We propose introducing immutability in backup storage by leveraging cloud provider features that support a write-once-read-many (WORM) model. This approach will prevent data alterations post-creation, enhancing data integrity and security.
+
+There are two types of immutability options to consider:
+
+1. **Bucket-Level Immutability:** Applies a uniform immutability period to all objects within a bucket. This is widely supported and easier to manage across different cloud providers.
+2. **Object-Level Immutability:** Allows setting immutability periods on a per-object basis, offering more granular control but with increased complexity and varying support across providers.
+
+In the detailed design, we will focus on bucket-level immutability policies due to their broader support and simpler management.
+
+### Detailed Design
+
+#### Bucket Immutability Mechanism
+
+The bucket immutability feature configures an immutability policy for a cloud storage bucket, dictating how long objects in the bucket must remain immutable. It also allows for locking the bucket's immutability policy, permanently preventing the policy from being reduced or removed.
```

Comment:
Unfortunately this would not be accurate. Cloud providers like Google Cloud Storage, and Azure Blob Storage **explicitly allow increasing** the immutability policy duration, even when the bucket is locked.

Consumers can always increase this duration if the duration they set in the beginning is not sufficient.